esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -6118,7 +6104,289 @@ bool CWsDfuEx::onDFUFileAccess(IEspContext &context, IEspDFUFileAccessRequest &r
+
+        getFileAccessBeforePublish(context, tempFileName.str(), jobId.str(), layoutBin, fileDesc, userDesc, requests, resp);
+
+        //create FileId
+        groupName.append(DFUFileIdSeparator).append(clusterName).append(DFUFileIdSeparator).append(tempFileName.str());
+        resp.setFileId(groupName.str());
+        ESPLOG(LogMax, "DFUFileCreate fileId:%s", groupName.str());
+    }
+    catch (IException *e)
+    {
+        FORWARDEXCEPTION(context, e,  ECLWATCH_INTERNAL_ERROR);
+    }
+    return true;
+}
+
+bool CWsDfuEx::onDFUFilePublish(IEspContext &context, IEspDFUFilePublishRequest &req, IEspDFUFilePublishResponse &resp)
+{
+    try
+    {
+        context.ensureFeatureAccess(FEATURE_URL, SecAccess_Write, ECLWATCH_DFU_ACCESS_DENIED, "WsDfu::DFUFilePublish: Permission denied.");
+
+        const char *fileId = req.getFileId();
+        const char *recordDefinition = req.getECLRecordDefinition();
+        if (isEmptyString(fileId))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT, "No FileId defined.");
+        if (isEmptyString(recordDefinition))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT, "No ECLRecordDefinition defined.");
+
+        StringBuffer userId, oldFileName, newFileName, clusterName, groupName;
+        const char *pGroupName = strchr(fileId, DFUFileIdSeparator);
+        if (!pGroupName)
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Failed to read Group Name from FileId '%s'.", fileId);
+
+        groupName.append(pGroupName - fileId, fileId);
+        const char *pClusterName = strchr(++pGroupName, DFUFileIdSeparator);
+        if (!pClusterName)
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Failed to read Cluster Name from FileId '%s'.", fileId);
+        clusterName.append(pClusterName - pGroupName, pGroupName);
+
+        oldFileName.set(++pClusterName);
+        newFileName.set(pClusterName);
+        if (newFileName.length() < 17)
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Invalid FileId: cannot read FileName.");
+        newFileName.setLength(newFileName.length() - 16); //".wsdfucreate.tmp"
+
+        Owned<IConstWUClusterInfo> clusterInfo = getTargetClusterInfo(clusterName.str());
+        ClusterType clusterType = clusterInfo->getPlatform();
+        if (clusterType == RoxieCluster)
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Roxie cluster is not allowed: %s.", clusterName.str());
+
+        MemoryBuffer layoutBin;
+        exportRecordDefinitionBinaryType(recordDefinition, layoutBin);
+
+        Owned<IUserDescriptor> userDesc;
+        context.getUserID(userId);
+        if (!userId.isEmpty())
+        {
+            userDesc.setown(createUserDescriptor());
+            userDesc->set(userId.str(), context.queryPassword(), context.querySignature());
+        }
+
+        VStringBuffer jobId("Publish %s on %s", newFileName.str(), clusterName.str());
+        ESPLOG(LogMax, "%s", jobId.str());
```

Comment:
This (and anywhere in esp that deals with Dali files) should really use : LOG(daliAuditLogCat,",FileAccess,
see thmfilemanager.cpp for examples.